pages/en/lb4/Creating-components.md

```diff
@@ -307,3 +307,297 @@ class EmailComponent {
+
+No matter what protocol you intend to use for your custom server, you'll need
+to use some algorithm to determine _which_ controller and function to send
+request data to, and that means you need a router.
+
+For example, consider a "toy protocol" similar to the JSON RPC
+specification (but nowhere near as complete or robust).
+
+The toy protocol will require a JSON payload with three properties: `controller`,
+`method`, and `input`.
+
+An example request would look something like this:
+```json
+{
+  "controller": "GreetController",
+  "method": "basicHello",
+  "input": {
+    "name": "world",
+  }
+}
+```
+
+### Defining the controller
+Here is the `GreetController` and the `Person` type:
+
+{% include code-caption.html content="/src/controllers/greet.controller.ts" %}
+
+```ts
+export class GreetController {
+  basicHello(input: Person) {
+    return `Hello, ${input && input.name || 'World'}!`;
+  }
+
+  hobbyHello(input: Person) {
+    return `${this.basicHello(input)} I heard you like ${input && input.hobby ||
+      'underwater basket weaving'}.`;
+  }
+}
+```
+
+{% include code-caption.html content="/src/models/person.model.ts" %}
+```ts
+// Note that this can also be a class!
+export type Person = {
+  name: string;
+  hobby: string;
+};
+```
+
+### Defining our router and server
+Now that there is a controller to test with, the following code creates a `Router` and `Server`.
+This example uses [express](https://expressjs.com/) and the
+`@types/express` package.
+
+{% include code-caption.html content="/src/server/rpc-router.ts" %}
+```ts
+import {RPCServer} from './rpc-server';
+import * as express from 'express';
+import * as parser from 'body-parser';
+
+export class RPCRouter {
+  routing: express.Router;
+  constructor(private server: RPCServer) {
+    this.routing = express.Router();
+    const jsonParser = parser.json();
+    this.routing.post('*', jsonParser, async (request, response) => {
+      console.log(JSON.stringify(request.body));
+      // Aliasing these values for clarity and brevity.
+      const ctrl = request.body.controller;
+      const method = request.body.method;
+      const input = request.body.input;
+      try {
+        const controller = await this.server.get(`controllers.${ctrl}`);
+        if (!controller) {
+          response.statusCode = 400;
+          response.send(`No controller was found with name "${ctrl}".`);
+          return;
+        }
+        if (!controller[method]) {
+          response.statusCode = 400;
+          response.send(
+            `No method was found on controller "${ctrl}" with name "${method}".`
+          );
+        }
+        response.send(await controller[method](input));
+      } catch (err) {
+        // You'll probably want more robust error handling in a real server!
+        console.error(err);
+        response.statusCode = 500;
+        response.send(err);
+      }
+    });
+    // Use our router!
+    this.server.expressServer.use(this.routing);
```

Comment:
Why is this custom router a class when all it does is to create an express router and mount it on the server? I find this design confusing. Also why are you creating a new Router instance, when there is only a single route set up? 

Here is a possibly simpler solution:

```ts
export function mountRpcRoute(server) {
  server.expressServer.post('*', jsonParser, async (request, response) => {
    handleRequest(server, request, response)
        .catch(err => this.handleError(server, request, response, error));
  });
}
```

Reply:
It was mostly to draw attention to the concept by isolating it as a class.

Reply:
Additionally, it makes it way simpler to mock pieces of the code if it's a class instead of a function.